Add HotKeys Page to DevTools

stevemonaco opened the pull request:
## What does the pull request do?

Fixes #15390 

Adds a HotKey page because there's no in-app discoverability of DevTools hotkeys. Otherwise, the user must check docs.

## What is the updated/expected behavior with this PR?

![devtools-hotkeys](https://github.com/AvaloniaUI/Avalonia/assets/1782158/ba7131dd-4fc6-4472-90cb-362e0930f11f)

Simple display of hotkeys, no mutability. Does not display the launch hotkey as that is mutable via `DevToolOptions` and would require some wiring. If the user doesn't know that hotkey, then how are they in DevTools to begin with?

Very recently, the "Styles" snapshot action was modified to "Value Frames". I reflect this in the hotkey descriptions, but other places still use Styles, including the DevTools docs page.

## How was the solution implemented (if it's not obvious)?

Adds an invisible `TabStripItem` for the `HotKeys` page to avoid visual crowding and avoid adding more complex navigation.

## Checklist

- [ ] Added unit tests (if possible)?
- [ ] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/avalonia-docs with user documentation

stevemonaco committed:
Add HotKeys Page to DevTools

stevemonaco commented on src/Avalonia.Diagnostics/Diagnostics/ViewModels/MainViewModel.cs:
```diff
@@ -231,6 +236,11 @@ private set
 
+        public void ShowHotKeys()
+        {
+            SelectedTab = 3;
```
A bit cludgy here, but changing `SelectedTab` is necessary so other tabs do not remain selected.

maxkatz6 commented on src/Avalonia.Diagnostics/Diagnostics/ViewModels/HotKeyPageViewModel.cs:
```diff
@@ -0,0 +1,21 @@
+    {
+        HotKeys = new()
+        {
+            new("Enable Snapshot Frames", "Pauses refreshing the Value Frames inspector for the selected Control", "Alt+S"),
+            new("Disable Snapshot Frames", "Resumes refreshing the Value Frames inspector for the selected Control", "Alt+D"),
+            new("Inspect Control Under Pointer", "Inspects the hovered Control in the Logical or Visual Tree Page", "Ctrl+Shift"),
+            new("Toggle Popup Freeze", "Prevents visible Popups from closing so they can be inspected", "Ctrl+Alt+F"),
+            new("Screenshot Selected Control", "Saves a Screenshot of the Selected Control in the Logical or Visual Tree Page", "F8")
```
This PR looks good to me, but it would be a nice improvement to define KeyGestures in one place, and reuse in event handling and these descriptions.

stevemonaco replied:
I'll amend to lift these into `DevToolsOptions` with some wiring work. It won't support mutability at runtime, but devs could provide their own KeyGestures at attach-time, read from their own config if they really wanted to.

maxkatz6 replied:
I didn't mean configurability, but only a static class with pre-defined key gestures. At least for now.
Long term, these should be mutable via devtools options, yeah.

stevemonaco committed:
Centralize hotkeys and hoist into DevToolsOptions

stevemonaco commented:
0761e382ba36af98826f04f96767a9f33b3386fb introduces `HotKeyConfiguration` to hold `KeyGesture`s as named members. It doesn't include `DevToolsOptions.Gesture` which launches DevTools as moving that is a breaking change.

I added `HotKeyConfiguration` to `DevToolsOptions`, use it to detect keypresses in `MainView`, moved the functionality into separate methods, and wire it into `HotKeyPageViewModel`. I can still list the launch DevTools gesture from there.

It's much closer to mutability, but kept it internal so any implementer could revise. This PR doesn't support multiple gestures that can each launch the same action. Chrome DevTools has this.

workgroupengineering commented:
possible enhanced

```c#
public enum HotKeyFunction
{
    [Description("Launch DevTools")]
    LaunchDevTools,
    [Description("Inspect Control Under Pointer")]
    InspectHoveredControl,
    [Description("Freeze Value Frames")]
    FramesFreeze,
    [Description("Unfreeze Value Frames")]
    FramesUnfreeze,
    [Description("Toggle Popup Freeze")]
    TogglePopupFreeze,
    [Description("Screenshot Selected Control")]
    ScreenshotSelectedControl
}

public class DevToolsOptions
{
    private readonly static KeyGesture DefaultLaunchDevToolsGesture = new(Key.F12);
    private readonly Dictionary<HotKeyFunction, KeyGesture> _hotkeyConfig = new()
    {
        {HotKeyFunction.LaunchDevTools,new(Key.F12) },
        {HotKeyFunction.InspectHoveredControl,new(Key.None, KeyModifiers.Shift | KeyModifiers.Control) },
        {HotKeyFunction.FramesFreeze,new(Key.S, KeyModifiers.Alt) },
        {HotKeyFunction.FramesUnfreeze,new(Key.D, KeyModifiers.Alt) },
        {HotKeyFunction.TogglePopupFreeze,new(Key.F, KeyModifiers.Alt | KeyModifiers.Control) },
        {HotKeyFunction.ScreenshotSelectedControl,new(Key.F8) },
    };



    /// <summary>
    /// Gets or sets the key gesture used to open DevTools.
    /// </summary>
    public KeyGesture Gesture
    {
        get => _hotkeyConfig.TryGetValue(HotKeyFunction.LaunchDevTools, out var gesture)
            ? gesture
            : DefaultLaunchDevToolsGesture;
        set => _hotkeyConfig[HotKeyFunction.LaunchDevTools] = value;
    }

    public IReadOnlyDictionary<HotKeyFunction, KeyGesture> HotKeyConfig
    {
        get => _hotkeyConfig;
        init
        {
            // Merge
            foreach (var kvp in value)
            {
                _hotkeyConfig[kvp.Key] = kvp.Value;
            }
        }
    }
...

}
``` 
using like:

```c#
var options = new DevToolsOptions()
{
    HotKeyConfig = new Dictionary<HotKeyFunction, KeyGesture>
    {
        { HotKeyFunction.LaunchDevTools , new(Key.F24) },
        { HotKeyFunction.InspectHoveredControl, new(Key.None, KeyModifiers.Shift | KeyModifiers.Control) },
    },
};
```

maxkatz6 commented:
We probably won't have this much of hotkeys to justify a dictionary. A class looks more than enough here.
Either way, it can be changed later while it's internal.
